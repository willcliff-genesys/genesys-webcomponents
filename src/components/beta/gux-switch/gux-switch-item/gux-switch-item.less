@import (reference) '../../../../style/color.less';
@import (reference) '../../../../style/spacing.less';
@import (reference) '../../../../style/typography.less';

@icon-off: #98a7b8;
@icon-on: #444a52;
@hover-on: #2a60c8;
@inactive-small: #e4e9f0;
@active-tab: #75a8ff;

gux-switch-item {
  > .gux-switch-item {
    .body-font();

    background: none;
    border: none;
    text-align: center;
    font: unset;
    color: @gux-type;
    outline: none;

    > gux-icon {
      svg {
        fill: @icon-off;
      }
    }
  }

  &[disabled] {
    opacity: 0.5;
    pointer-events: none;
  }

  &.gux-selected {
    .bold-font();
  }
}

gux-switch-beta {
  // Default
  > gux-switch-item {
    .gux-switch-item {
      border-bottom: 1px solid @inactive-small;
      padding: 0 @spacing-xs @spacing-2xs;

      &:focus,
      &:hover {
        outline: none;
        border-bottom: @spacing-2xs solid @gux-dark-blue;
        padding: 0 @spacing-xs 1px;

        > gux-icon {
          svg {
            fill: @icon-on;
          }
        }
      }

      > gux-icon {
        height: 32px;
        width: 32px;
      }
    }

    &.gux-selected {
      .gux-switch-item {
        border-bottom: @spacing-2xs solid @gux-dark-blue;
        padding: 0 @spacing-xs 1px;

        > gux-icon {
          svg {
            fill: @icon-on;
          }
        }
      }
    }
  }

  &.gux-small {
    > gux-switch-item {
      .gux-switch-item {
        .gux-small-text();

        border-bottom: 1px solid @inactive-small;
        padding: 0 @spacing-2xs;

        &:focus,
        &:hover {
          border-bottom-color: @icon-on;
        }

        > gux-icon {
          height: 24px;
          width: 24px;
        }
      }

      &.gux-selected {
        .gux-switch-item {
          border-bottom-color: @icon-on;
        }
      }
    }
  }
}
