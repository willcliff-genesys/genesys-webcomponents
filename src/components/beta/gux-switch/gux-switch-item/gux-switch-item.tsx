import { Component, h, Host, Listen, JSX, Prop } from '@stencil/core';

@Component({
  styleUrl: 'gux-switch-item.less',
  tag: 'gux-switch-item'
})
export class GuxSwitchitem {
  @Prop()
  value: string;

  @Prop()
  disabled: boolean = false;

  @Listen('click')
  onClick(e: MouseEvent): void {
    if (this.disabled) {
      e.stopPropagation();
    }
  }

  render(): JSX.Element {
    return (
      <Host>
        <button type="button" class="gux-switch-item" disabled={this.disabled}>
          <slot />
        </button>
      </Host>
    );
  }
}
