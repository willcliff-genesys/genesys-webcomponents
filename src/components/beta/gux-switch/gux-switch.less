gux-switch-beta {
  display: flex;
  align-items: flex-end;
}
