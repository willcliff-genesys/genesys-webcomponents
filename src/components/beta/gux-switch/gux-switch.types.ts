export type AllowedLayouts = 'small' | 'default';
