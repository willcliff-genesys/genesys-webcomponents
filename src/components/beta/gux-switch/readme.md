# gux-switch

## Native Events

| Event    | Description | Type               |
| -------- | ----------- | ------------------ |
| `change` |             | `InputEvent`       |
| `input`  |             | `InputEvent`       |

<!-- Auto Generated Below -->


## Properties

| Property | Attribute | Description                                        | Type                   | Default     |
| -------- | --------- | -------------------------------------------------- | ---------------------- | ----------- |
| `layout` | `layout`  | The allowed sizes                                  | `"default" \| "small"` | `'default'` |
| `value`  | `value`   | Used to keep track of the currently selected value | `string`               | `undefined` |


## Slots

| Slot | Description                      |
| ---- | -------------------------------- |
|      | list of gux-switch-item elements |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
