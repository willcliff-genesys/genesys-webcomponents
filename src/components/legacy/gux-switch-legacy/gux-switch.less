@import (reference) '../../../style/color.less';

// Variables part

// Style
.gux-switch-group {
  display: inline-block;
  padding: 0;

  .gux-switch {
    border: none;
    box-shadow: inset 0 -1px 0 0 @gux-border;
    display: inline-block;

    .switch-button {
      border: none;
      background: transparent;
      padding: 0;

      .display-text {
        font-size: 12px;
        padding: 8px 0;
        margin: 0 8px;
      }

      &:hover,
      &:active,
      &:focus {
        cursor: pointer;
        color: @gux-genesys-dark-blue;
      }

      &:focus {
        outline: none;
      }
    }

    &:first-child {
      .display-text {
        margin-left: 0;
      }
    }

    &:last-child {
      .display-text {
        margin-right: 0;
      }
    }

    &.selected {
      .display-text {
        box-shadow: inset 0 -1px 0 0 @gux-charcoal-grey;
        font-weight: bold;
      }
    }

    &.small {
      .display-text {
        font-size: 10px;
      }
    }

    &.large {
      .display-text {
        font-size: 14px;
      }

      &.selected {
        .display-text {
          box-shadow: inset 0 -4px 0 0 @gux-genesys-blue;
        }
      }
    }

    &:hover,
    &:active {
      color: @gux-genesys-dark-blue;
    }

    &.disabled {
      opacity: 0.5;
      pointer-events: none;
    }
  }
}

// Theming

// Dark
.gux-switch-dark-theme {
  .gux-switch {
    box-shadow: inset 0 -1px 0 0 @gux-grey-global;

    .switch-button {
      .display-text {
        color: @gux-type-dark;
      }

      &:hover,
      &:active {
        cursor: pointer;

        .display-text {
          color: @gux-genesys-dark-blue;
        }
      }
    }

    &.selected {
      .display-text {
        box-shadow: inset 0 -1px 0 0 @gux-genesys-blue;
        font-weight: bold;
      }

      &.large {
        .display-text {
          box-shadow: inset 0 -4px 0 0 @gux-genesys-blue;
        }
      }
    }
  }
}

.gux-dark-theme {
  .gux-switch-group {
    .gux-switch-dark-theme();
  }
}

.gux-switch-group.gux-dark-theme {
  .gux-switch-dark-theme();
}

// Light
.gux-switch-light-theme {
  color: @gux-type;
}

.gux-light-theme {
  .gux-switch-group {
    .gux-switch-light-theme();
  }
}

.gux-switch-group.gux-light-theme {
  .gux-switch-light-theme();
}

// Default Theme
.gux-switch-group {
  .gux-switch-light-theme();
}
