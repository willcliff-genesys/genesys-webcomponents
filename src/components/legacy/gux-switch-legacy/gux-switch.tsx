import { Component, Event, EventEmitter, h, Prop } from '@stencil/core';
import { JSXBase } from '@stencil/core/internal';
import { KeyCode } from '../../../common-enums';

export type AllowedLayouts = 'small' | 'medium' | 'large';

export interface ISwitchItem {
  displayName: string;
  isDisabled?: boolean;
  value: string;
}

@Component({
  styleUrl: 'gux-switch.less',
  tag: 'gux-switch-legacy'
})
export class GuxSwitch {
  /**
   * The names and values of the switch buttons
   */
  @Prop()
  items: ISwitchItem[];

  /**
   * Used to keep track of the currently selected value
   */
  @Prop({ mutable: true })
  selectedValue: string;

  /**
   * The allowed sizes
   */
  @Prop()
  layout: AllowedLayouts = 'medium';

  /**
   * Triggers when a switch is selected.
   * @return the value of the newly selected switch
   */
  @Event()
  selectionChanged: EventEmitter;

  selectSwitchAtIndex(index: number) {
    this.selectedValue = this.items[index].value;
    this.selectionChanged.emit(this.selectedValue);
  }

  onKeyDownAtIndex(index: number, event: KeyboardEvent) {
    if (this.items[index].isDisabled) {
      event.stopPropagation();
    } else if ([KeyCode.Enter, KeyCode.Space].includes(event.keyCode)) {
      this.selectSwitchAtIndex(index);
    }
  }

  classForSwtichItemAtIndex(item: ISwitchItem): string {
    return `gux-switch ${this.layout}${
      item.value === this.selectedValue ? ' selected' : ''
    }${item.isDisabled ? ' disabled' : ''}`;
  }

  listElementForSwtichItemAtIndex(
    item: ISwitchItem,
    index: number
  ): JSXBase.LiHTMLAttributes<HTMLLIElement> {
    if (item.isDisabled) {
      return (
        <li class={this.classForSwtichItemAtIndex(item)}>
          <button class="switch-button" disabled={item.isDisabled}>
            <div class="display-text">{item.displayName}</div>
          </button>
        </li>
      );
    }
    return (
      <li class={this.classForSwtichItemAtIndex(item)}>
        <button
          class="switch-button"
          disabled={item.isDisabled}
          onKeyDown={ev => this.onKeyDownAtIndex(index, ev)}
          onClick={() => this.selectSwitchAtIndex(index)}
        >
          <div class="display-text">{item.displayName}</div>
        </button>
      </li>
    );
  }

  render() {
    return (
      <ul class="gux-switch-group">
        {this.items &&
          this.items.map((item, index) =>
            this.listElementForSwtichItemAtIndex(item, index)
          )}
      </ul>
    );
  }
}
