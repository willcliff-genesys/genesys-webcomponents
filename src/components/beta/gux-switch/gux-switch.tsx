import { Component, Element, h, Host, Listen, JSX, Prop } from '@stencil/core';

import { AllowedLayouts } from './gux-switch.types';

import simulateNativeEvent from '../../../utils/simulate-native-event/simulate-native-event';

/**
 * @slot - list of gux-switch-item elements
 */
@Component({
  styleUrl: 'gux-switch.less',
  tag: 'gux-switch-beta'
})
export class GuxSwitch {
  @Element() root: HTMLElement;

  /**
   * Used to keep track of the currently selected value
   */
  @Prop({ mutable: true })
  value: string;

  /**
   * The allowed sizes
   */
  @Prop()
  layout: AllowedLayouts = 'default';

  @Listen('click')
  onClick(e: MouseEvent): void {
    e.stopPropagation();

    const switchItem = (e.target as HTMLElement).closest('gux-switch-item');

    if (switchItem && this.value !== switchItem.value) {
      this.value = switchItem.value;

      simulateNativeEvent(this.root, 'input');
      simulateNativeEvent(this.root, 'change');
    }
  }

  private updateSelectedItem(switchItems: HTMLGuxSwitchItemElement[]): void {
    switchItems.forEach(switchItem => {
      if (switchItem.value === this.value) {
        switchItem.classList.add('gux-selected');
      } else {
        switchItem.classList.remove('gux-selected');
      }
    });
  }

  render(): JSX.Element {
    this.updateSelectedItem(
      Array.from(this.root.children) as HTMLGuxSwitchItemElement[]
    );

    return (
      <Host role="group" class={`gux-${this.layout}`}>
        <slot />
      </Host>
    );
  }
}
