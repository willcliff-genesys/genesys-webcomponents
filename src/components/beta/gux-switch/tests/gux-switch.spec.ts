import { newSpecPage } from '@stencil/core/testing';
import '../../../../components';

import { GuxSwitch } from '../gux-switch';
import { GuxSwitchitem } from '../gux-switch-item/gux-switch-item';

global.InputEvent = Event;

const components = [GuxSwitch, GuxSwitchitem];
const html = `
  <gux-switch-beta layout="small" value="day">
    <gux-switch-item value="month">Month</gux-switch-item>
    <gux-switch-item value="week">Week</gux-switch-item>
    <gux-switch-item value="day">Day</gux-switch-item>
    <gux-switch-item value="hour" disabled>Hour</gux-switch-item>
    <gux-switch-item value="minute">Minute</gux-switch-item>
  </gux-switch-beta>
`;
const language = 'en';

describe('gux-switch', () => {
  describe('#render', () => {
    it(`should render as expected`, async () => {
      const page = await newSpecPage({ components, html, language });

      expect(page.rootInstance).toBeInstanceOf(GuxSwitch);
      expect(page.root).toMatchSnapshot();
    });
  });

  describe('#interactions', () => {
    it(`should change value on gux-switch-item click`, async () => {
      const page = await newSpecPage({ components, html, language });
      const element = page.root as HTMLGuxSwitchBetaElement;
      const guxSwitchItemMinute = page.root.querySelector(
        'gux-switch-item[value=minute]'
      ) as HTMLGuxSwitchItemElement;

      expect(element.value).toBe('day');

      guxSwitchItemMinute.click();
      await page.waitForChanges();

      expect(element.value).toBe(guxSwitchItemMinute.value);
    });

    it(`should not change value on gux-switch-item click if it is disabled`, async () => {
      const page = await newSpecPage({ components, html, language });
      const element = page.root as HTMLGuxSwitchBetaElement;
      const guxSwitchItemHour = page.root.querySelector(
        'gux-switch-item[value=hour]'
      ) as HTMLGuxSwitchItemElement;
      const currentValue = element.value;

      expect(currentValue).toBe('day');

      guxSwitchItemHour.click();
      await page.waitForChanges();

      expect(element.value).toBe(currentValue);
    });

    it(`should emit a 'change' and 'input' event when a new item is selected`, async () => {
      const page = await newSpecPage({ components, html, language });
      const element = page.root as HTMLGuxSwitchBetaElement;
      const guxSwitchItemMinute = page.root.querySelector(
        'gux-switch-item[value=minute]'
      ) as HTMLGuxSwitchItemElement;
      const changeEventSpy = jest.fn();
      const inputEventSpy = jest.fn();

      element.addEventListener('change', () => {
        changeEventSpy();
      });
      element.addEventListener('input', () => {
        inputEventSpy();
      });

      guxSwitchItemMinute.click();
      await page.waitForChanges();

      expect(changeEventSpy).toHaveBeenCalledWith();
      expect(inputEventSpy).toHaveBeenCalledWith();
    });

    it(`should not emit a 'change' or 'input' event when a disabled item is selected`, async () => {
      const page = await newSpecPage({ components, html, language });
      const element = page.root as HTMLGuxSwitchBetaElement;
      const guxSwitchItemHour = page.root.querySelector(
        'gux-switch-item[value=hour]'
      ) as HTMLGuxSwitchItemElement;
      const changeEventSpy = jest.fn();
      const inputEventSpy = jest.fn();

      element.addEventListener('change', () => {
        console.info('change');
        changeEventSpy();
      });
      element.addEventListener('input', () => {
        console.info('change');
        inputEventSpy();
      });

      guxSwitchItemHour.click();
      await page.waitForChanges();

      expect(changeEventSpy).not.toHaveBeenCalled();
      expect(inputEventSpy).not.toHaveBeenCalled();
    });
  });
});
