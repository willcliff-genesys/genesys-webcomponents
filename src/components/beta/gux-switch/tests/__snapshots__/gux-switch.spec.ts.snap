// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-switch #render should render as expected 1`] = `
<gux-switch-beta class="gux-small" layout="small" role="group" value="day">
  <!---->
  <gux-switch-item value="month">
    <!---->
    <button class="gux-switch-item" type="button">
      Month
    </button>
  </gux-switch-item>
  <gux-switch-item value="week">
    <!---->
    <button class="gux-switch-item" type="button">
      Week
    </button>
  </gux-switch-item>
  <gux-switch-item class="gux-selected" value="day">
    <!---->
    <button class="gux-switch-item" type="button">
      Day
    </button>
  </gux-switch-item>
  <gux-switch-item disabled="" value="hour">
    <!---->
    <button class="gux-switch-item" disabled="" type="button">
      Hour
    </button>
  </gux-switch-item>
  <gux-switch-item value="minute">
    <!---->
    <button class="gux-switch-item" type="button">
      Minute
    </button>
  </gux-switch-item>
</gux-switch-beta>
`;
